### ISSUE TITLE
unified/side-by-side diff controls are confusing

### ISSUE THREAD

**Issue by erikwright** [Reviewable/Reviewable#324]
The "single/auto/<->" controls were not obviously related to this at all, even after I noticed that they were there. I still don't understand why the column setting changes to <-> when you're in auto mode (and all this button appears to do is turn off auto).

It's really annoying to pull the column control from 80 columns to 120. You have to drag it out to the edge of the screen to get half-way, and then let the page resize, and then do it again.

Suggestions:

(1) Make the word "diff" appear somewhere in these controls so it's obvious what they are for ("single", out of context, is meaningless).
(2) I think "unified" and "side-by-side" are well-recognized names for these diff styles. "unified diff" as a button would have made it immediately clear what the controls are for.
(3) Instead of having two toggles (for "single/(split)" and "auto/(manual)") that only actually have three valid combined states, consider a drop-down for "single/split/auto" (or, of course, "unified/side-by-side/auto").
(4) For the column limit, consider having a drop-down with a few common options (presumably 80 and 120) plus a custom value option.

**Comment by pkaminski** [comment]
I agree that the current widget has outgrown its original design, but I haven't figured out quite what to do with it yet -- whether to leave it where it is and improve it, or change to a completely different design.  Here's how it works for now:

- In 'auto' mode, Reviewable will dynamically pick the number of columns to fit the window width.  If 'single' is off, it will first try to fit side-by-side (down to 80 columns each), then fall back to unified diff.
- If 'auto' is off (whether because it was switched off manually or because you dragged the margin notch), the number of columns is hard-set to the number displayed.  If 'single' is off, side-by-side with that number of columns is tried for fit first, otherwise falls back to unified diff.
- If 'single' is on, side-by-side is precluded and unified diff is forced in all cases.

The basic effect is that if you've set a number of columns, then that's prioritized over a side-by-side view.  If you've set 'auto' then side-by-side is prioritized when it fits, even if the number of columns has to be reduced to 80.  If you've set 'single' then side-by-side is never attempted.

To your specific points:
(1) and (2) I don't believe there's widespread agreement on names for the two display styles, but GitHub uses "unified" and "split" so I should probably copy that.
(3) There are actually 4 distinct states.
(4) I agree that the margin-dragging hasn't worked out great in practice and should probably be replaced with a simple form field.

Since the margin notch is difficult to find and the space limitations make it hard to place a clear UI there, I might just move these settings to the account drop-down.

**Comment by erikwright** [comment]
On Thu, Jun 2, 2016 at 9:50 PM, Piotr Kaminski <notifications@github.com>
wrote:

> I agree that the current widget has outgrown its original design, but I
> haven't figured out quite what to do with it yet -- whether to leave it
> where it is and improve it, or change to a completely different design.
> Here's how it works for now:
>
>    - In 'auto' mode, Reviewable will dynamically pick the number of
>    columns to fit the window width. If 'single' is off, it will first try to
>    fit side-by-side (down to 80 columns each), then fall back to unified diff.
>    - If 'auto' is off (whether because it was switched off manually or
>    because you dragged the margin notch), the number of columns is hard-set to
>    the number displayed. If 'single' is off, side-by-side with that number of
>    columns is tried for fit first, otherwise falls back to unified diff.
>    - If 'single' is on, side-by-side is precluded and unified diff is
>    forced in all cases.
>
> The basic effect is that if you've set a number of columns, then that's
> prioritized over a side-by-side view. If you've set 'auto' then
> side-by-side is prioritized when it fits, even if the number of columns has
> to be reduced to 80. If you've set 'single' then side-by-side is never
> attempted.
>
> To your specific points:
> (1) and (2) I don't believe there's widespread agreement on names for the
> two display styles, but GitHub uses "unified" and "split" so I should
> probably copy that.
> (3) There are actually 4 distinct states.
> (4) I agree that the margin-dragging hasn't worked out great in practice
> and should probably be replaced with a simple form field.
>
> Since the margin notch is difficult to find and the space limitations make
> it hard to place a clear UI there, I might just move these settings to the
> account drop-down.
>
In larger organizations it _can_ be useful to have different column limits.
For example, in Chrome team there is an 80-char limit for almost
everything, but for Java files (IIRC) it's 120.

I think you treat the column count as a feature of the view, but in fact it
is a feature of the code. It's the maximum line length of what I'm
reviewing, and the purpose of being aware of that is that my viewer should
not wrap anything shorter than that and (ideally) should visually indicate
(by wrapping, highlighting, or with a vertical rule) anything that exceeds
it.

My personal opinion is that Reviewable should just respect what I ask it to
do. If I say "single", give me a unified diff. Otherwise, side-by-side.
Even if it overflows the screen. Either overflow the view area or introduce
horizontal scrolling in the diff area. "Auto" is a fine default, but I
think it should just mean "choose the best style (unified/side-by-side)
according to my column count and window size". It's a mistake to treat the
column count as subject to automatic change.

**Comment by pkaminski** [comment]
Opinion noted, but an equally valid (and fairly popular) approach is to not have stated line length limits for code, in which case fitting layout to available space is ideal.  In the absence of information about max line length this is also a sensible default.

I strongly disagree that overflowing / side-scrolling is ever the right thing to do.  It makes for terrible UX, and makes it unnecessarily annoying to switch between devices (e.g., from a high-res desktop screen to a tablet).  So when push comes to shove and declared preferences result in an overflowing layout, I chose to prioritize the line length and switch to a unidiff.

I agree that line length limits are often a property of the code and I'm considering moving them into the admin-driven repo config.  I think language-specific line lengths are a pretty rare phenomenon, though -- I haven't encountered this particular perversion outside Google yet.  I do also think that sometimes a user may want to override the project's line length, either because it doesn't have one set, because it shouldn't apply to a specific file for some reason, or because they'd rather have a split diff even if it means wrapping some lines.

I'm not yet sure how to reform the settings to account for the above without making them overly complex, though.

**Comment by erikwright** [comment]
On Mon, Jun 6, 2016 at 12:54 AM, Piotr Kaminski <notifications@github.com>
wrote:

> Opinion noted, but an equally valid (and fairly popular) approach is to
> not have stated line length limits for code, in which case fitting layout
> to available space is ideal. In the absence of information about max line
> length this is also a sensible default.
>
> I strongly disagree that overflowing / side-scrolling is ever the right
> thing to do. It makes for terrible UX, and makes it unnecessarily annoying
> to switch between devices (e.g., from a high-res desktop screen to a
> tablet). So when push comes to shove and declared preferences result in an
> overflowing layout, I chose to prioritize the line length and switch to a
> unidiff.
>
> I agree that line length limits are often a property of the code and I'm
> considering moving them into the admin-driven repo config. I think
> language-specific line lengths are a pretty rare phenomenon, though -- I
> haven't encountered this particular perversion outside Google yet. I do
> also think that sometimes a user may want to override the project's line
> length, either because it doesn't have one set, because it shouldn't apply
> to a specific file for some reason, or because they'd rather have a split
> diff even if it means wrapping some lines.
>
> I'm not yet sure how to reform the settings to account for the above
> without making them overly complex, though.
>
Ultimately, if the controls were simpler, I think almost any behaviour
would be OK and the developer would just do what is required (adjust line
lengths, change their window size, deal with a unified diff, whatever). The
problem is right now that the effect of the controls is incredibly
confusing so the developer cannot even tell why they aren't getting what
they want or what is required in order to do it.

**Comment by kballard** [comment]
> I think language-specific line lengths are a pretty rare phenomenon, though -- I haven't encountered this particular perversion outside Google yet.

Rust code is usually wrapped at 100 characters. Python's recommended style is to wrap at 79 ([PEP-8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length)). I'm not sure offhand what other languages have recommended line lengths though.

> I do also think that sometimes a user may want to override the project's line length, either because it doesn't have one set, because it shouldn't apply to a specific file for some reason, or because they'd rather have a split diff even if it means wrapping some lines.

I agree.

**Comment by thockin** [comment]
Chiming in late - I just want to see it as wide as my screen.  I am using Go, which eschews line-length limits, but sometimes I look at markdown or other languages and I want limits.  I'd prefer to have an optional limit and denote when code goes over that limit, but NOT force a wrap at that limit.

E.g. I try to keep my go code under 100 columns, but sometimes not.  I want to see the full length of the line on one line.

**Comment by sseo289** [comment]
https://crackzero.com/ccleaner-professional-crack/
https://crackzero.com/ccleaner-professional-crack/

https://crackzero.com/ccleaner-professional-crack/

https://crackzero.com/ccleaner-professional-crack/
